### ISSUE TITLE
Remote include not working in the Maven plugin

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor-maven-plugin#153]
A remote include from GitHub is not working when using the Maven plugin. It's resolving the include to an empty string.

Here's an example:

```
[indent=0]
....
include::https://raw.githubusercontent.com/cometd/cometd/master/pom.xml[tag=jetty-version]
....
```

This works when using both the Asciidoctor and AsciidoctorJ commandline tools. For some reason, it fails in the Maven plugin. We should include a test for this so we can ensure it doesn't break once we fix it.

**Comment by mojavelinux** [comment]
This works in the Gradle plugin, so the problem seems to be with the Maven plugin or a conflict with Maven libraries. I remember @lordofthejars mentioning something about BouncyCastle perhaps?

**Comment by paranoiabla** [comment]
It's probably because the HTML generated from Maven is different. This is because maven-site-plugin uses Doxia which sanitizes the produced html and removes a lot of the attributes on different tags (like class, style, etc..)

**Comment by mojavelinux** [comment]
While that's true, this isn't related to that. The core functionality of Asciidoctor of including a file from a remote URL is failing in this case, but silently. I don't have time right now to debug, but I'm sure doing so would reveal what is happening...perhaps a swallowed exception.

**Comment by LightGuard** [comment]
I'd be surprised if BouncyCastle had anything to do with this, it's for
cryptography. There may be a conflict somewhere, but we're not doing
anything with the asciidoc before we send it over. Maybe it's the security
we're running under?

On Tue, Mar 31, 2015 at 9:05 AM, Dan Allen <notifications@github.com> wrote:

> While that's true, this isn't related to that. The core functionality of
> Asciidoctor of including a file from a remote URL is failing in this case,
> but silently. I don't have time right now to debug, but I'm sure doing so
> would reveal what is happening...perhaps a swallowed exception.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/asciidoctor/asciidoctor-maven-plugin/issues/153#issuecomment-88124490>
> .
>



-- 
Jason Porter
http://en.gravatar.com/lightguardjp

**Comment by mojavelinux** [comment]
I quickly tried a bunch of things. That's why I think getting it into the debugger is the next step because it's probably trying to tell us something, but it's getting swallowed.

**Comment by mojavelinux** [comment]
I'm still seeing this and it has me totally confused. I can't understand why Asciidoctor is failing to resolve the remote content when called from Maven. Something is different on the classpath than when we used AsciidcotorJ from all other contexts. Note that the issue is in Asciidoctor core itself, but Asciidoctor core is behaving different in this environment.

**Comment by mojavelinux** [comment]
What we need to do is inspect the HTTP traffic triggered by the following line in Asciidoctor core.

https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/reader.rb#L938

**Comment by mojavelinux** [comment]
I finally got to the bottom of it. The combination of two condition lead to this problem:

1. GitHub sends back compressed content by default (zlib or gzip, or whatever)
2. JRuby 1.7 running in 2.0 mode does not properly uncompress content

We switched to Ruby 2.0 mode in AsciidoctorJ 1.5.1, which is when this stopped working (in combination with GitHub's change).

There are four possible solutions, the first two of which are possible to control from the Maven configuration:

1. Use AsciidoctorJ 1.5.0 (obviously, not ideal)
2. Upgrade to JRuby 9.0.0.0.pre2 (I checked, it's a drop-in replacement, and correctly works in Ruby 2.0 mode)
3. Add the option `Accept-Encoding' => ''` where the `open` method is used in Asciidoctor core
4. Set JRuby back to 1.9 mode

The quickest solution is (2). That's what I recommend for @sbordet.

Add the following to the Maven configuration of the plugin:

```xml
<dependencies>
    <dependency>
        <groupId>org.jruby</groupId>
        <artifactId>jruby-complete</artifactId>
        <version>9.0.0.0.pre2</version>
    </dependency>
</dependencies>
```

The change I propose for the plugin (and thus AsciidoctorJ as well) is that it has to be simple to be able to set the JRuby mode (4). We can't hardcode it to 2.0 and make it impossible to change back. Whether we need a code or documentation change, we need to make this customization possible. Right now, it renders AsciidoctorJ unusable if you are running into a Ruby 2.0 mode bug like this one and need to switch back to 1.9 mode.

I'd rather not do (3).

**Comment by mojavelinux** [comment]
We definitely need a test for a remote URL in the Maven plugin. It's clearly not enough that we have coverage in Asciidoctor core.

**Comment by sbordet** [comment]
@mojavelinux you can point to CometD's remote URL at https://github.com/cometd/cometd/blob/master/pom.xml.

Lines may change, but I will keep the tag in there because it's needed by the project documentation anyway.

Workaround 2 works fine for me, thanks !

**Comment by mojavelinux** [comment]
Indeed!

**Comment by NicolasGeraud** [comment]
hi @mojavelinux I've add the following to my pom.xml but I can't figure out why I can't include an external link in my asciidoc file :
```xml
<plugin>
    <groupId>org.asciidoctor</groupId>
    <artifactId>asciidoctor-maven-plugin</artifactId>
    <version>1.5.2</version>
    <dependencies>
        <dependency>
            <groupId>org.jruby</groupId>
            <artifactId>jruby-complete</artifactId>
            <version>9.0.0.0.pre2</version>
        </dependency>
    </dependencies>
    <executions>
        <execution>
            <id>output-single-html</id>
            <phase>generate-resources</phase>
            <goals>
                <goal>process-asciidoc</goal>
            </goals>
            <configuration>
                <backend>html5</backend>
                <doctype>book</doctype>
                <outputDirectory>${project.build.directory}/generated-docs/${project.version}/reference/htmlsingle/</outputDirectory>
                <attributes>
                    <safe-mode>UNSAFE</safe-mode>
                    <allow-uri-read>true</allow-uri-read>
                </attributes>
            </configuration>
        </execution>
        ...
```

What I'm doing wrong in my maven config file ?

**Comment by abelsromero** [comment]
Just adding` <allow-uri-read>true</allow-uri-read>` is enough, mode is UNSAFE by default.
But I haven't been able to make a GitHub include to work, not even using the direct asciidoctor gem in CLI, both on MacOs and Linux. I always get the message `include uri not readable`l.
Some research brought me to this https://github.com/asciidoctor/asciidoctor/issues/1049, but I guess it should work on *nix systems right?

**Comment by mojavelinux** [comment]
If you use JRuby 9.0.0.0 it should work. I noticed the sample is using JRuby 9.0.0.0.pre2. The final version is now out, so the first thing I would do is check if it works with JRuby 9.0.0.0.

But the real issue is that AsciidoctorJ is setting the Ruby mode which causes JRuby to behave badly. That fix is planned (and already made) for AsciidoctorJ 1.6.0. I recommend trying out one of the snapshots.

http://oss.jfrog.org/simple/oss-snapshot-local/org/asciidoctor/asciidoctorj/1.6.0-SNAPSHOT/

If AsciidoctorJ 1.6.0 doesn't work, then we really need to focus on getting it right there. Unfortunately, AsciidoctorJ 1.5.x is just going to somewhat unpredictable in this area since it puts Ruby into a mode that makes it perform in an inconsistent/unstable way.

**Comment by NicolasGeraud** [comment]
OK, I will try and post feedbacks

**Comment by NicolasGeraud** [comment]
Hi all,

I confirm it works with asciidoctor-maven-plugin 1.5.3 and jruby-complete 9.0.0.0

**Comment by abelsromero** [comment]
> We should include a test for this so we can ensure it doesn't break once we fix it.

We can add a test that right now asserts the 'issue'. This will fails when we upgrade jruby with AsciidoctorJ 1.6.0, will serve as a reminder.
wdyt?

**Comment by NicolasGeraud** [comment]
:+1:

**Comment by abelsromero** [comment]
This has been fixed in release 1.5.5 after upgrading jRuby. I will close this.
